language/diem-framework/modules/0L/TowerState.move

```diff
@@ -521,6 +572,31 @@ module TowerState {
       false 
     }
 
+    public fun get_fullnode_proofs_in_epoch(): u64 acquires TowerCounter{
+      let state = borrow_global<TowerCounter>(CoreAddresses::VM_RESERVED_ADDRESS());
+      state.fullnode_proofs_in_epoch
+    }
+
+    public fun get_fullnode_proofs_in_epoch_above_thresh(): u64 acquires TowerCounter{
+      let state = borrow_global<TowerCounter>(CoreAddresses::VM_RESERVED_ADDRESS());
+      state.fullnode_proofs_in_epoch_above_thresh
+    }
+
+    public fun get_lifetime_proof_count(): (u64, u64, u64) acquires TowerCounter{
+      let s = borrow_global<TowerCounter>(CoreAddresses::VM_RESERVED_ADDRESS());
+      (s.lifetime_proofs, s.lifetime_validator_proofs, s.lifetime_fullnode_proofs)
+    }
+
+    public fun danger_migrate_get_lifetime_proof_count(): (u64, u64, u64) acquires TowerStats{
+      if (exists<TowerStats>(CoreAddresses::VM_RESERVED_ADDRESS())) {
```

Comment:
Should this destroy TowerStats once it is run?

Reply:
In general I would say yes. But in this case I think it should NOT destroy the old struct. Just so we have it on chain for debugging, since this is the first on-chain migration we are doing. In a subsequent migrations we could do that.

Good catch